controllers/custom/custom_controller.go

```diff
@@ -79,79 +68,39 @@ type Options struct {
 // In future, we may be able to switch to upstream controller for reconciling Pods if the
 // long term solutions are in place
 type CustomController struct {
-	// workQueue to store create/update/delete events
-	workQueue workqueue.RateLimitingInterface
-	// mu is the mutex to allow the controller to sync before
-	// other controller start
-	mu sync.Mutex
-	// log for custom controller
 	log logr.Logger
-	// Reconciler will be called on all the K8s object events
-	Do Reconciler
-	// config to create a new client-go controller
+	// Controller is the K8s Controller
+	controller cache.Controller
+	// dataStoreSynced when set to true indicates the data store is synced and ready to access
+	dataStoreSynced *bool
+	// config for creating low level controller
 	config *cache.Config
-	// options is the configurable parameters for creating
-	// the controller
-	options  Options
-	syncFlag *bool
-}
-
-// Request for Add/Update only contains the Namespace/Name
-// Request for Delete contains the Pod Object as by the time
-// Delete Request is reconciled the cache will not have it
-type Request struct {
-	// Add/Update Request will contain the Namespaced Name only. The
-	// item can be retrieved from the indexer for add/update events
-	NamespacedName types.NamespacedName
-	// Delete Event will contain the DeletedObject only.
-	DeletedObject interface{}
 }
 
 // Starts the low level controller
 func (c *CustomController) Start(stop <-chan struct{}) error {
-	// This is important to allow the data store to be synced
-	// Before the other controller starts
-	c.mu.Lock()
-	// Shut down the queue so the worker can stop
-	defer c.workQueue.ShutDown()
-
-	err := func() error {
-		defer c.mu.Unlock()
-
-		coreController := cache.New(c.config)
-
-		c.log.Info("starting custom controller")
-		go coreController.Run(stop)
+	c.log.Info("starting custom controller")
+	c.controller = cache.New(c.config)
 
-		// Wait till cache sync
-		c.WaitForCacheSync(coreController)
-
-		c.log.Info("Starting Workers", "worker count",
-			c.options.MaxConcurrentReconciles)
-		for i := 0; i < c.options.MaxConcurrentReconciles; i++ {
-			go wait.Until(c.worker, time.Second, stop)
-		}
-
-		return nil
+	go func() {
+		c.MarkDataStoreSynced()
+		c.log.Info("data store has synced")
 	}()
-	if err != nil {
-		return err
-	}
 
-	<-stop
-	c.log.Info("stopping workers")
+	// Run the controller
+	c.controller.Run(stop)
+
 	return nil
 }
 
-// WaitForCacheSync tills the cache has synced, this must be done under
-// mutex lock to prevent other controllers from starting at same time
-func (c *CustomController) WaitForCacheSync(controller cache.Controller) {
-	for !controller.HasSynced() && controller.LastSyncResourceVersion() == "" {
+// MarkDataStoreSynced sets the dataStoreSynced variable to true when controller has
```

Comment:
I hope there is a test for this. This seems like a very tricky function here catering to few situations like nil or waiting for sync.